DAPHNE decode updates

wesketchum opened the pull request:
This has #105 already included in it (so should be merged after that is approved...)

This is a PR to have some updates on the DAPHNE decoder, including printout of TP info in the DAPHNEFrames for debug purposes.

wesketchum committed:
updates to the daphne decoder, to allow printouts of TP info for the DAPHNEFrames

wesketchum committed:
Merge remote-tracking branch 'origin/patch/fddaq-v5.3.x' into wketchum/daphne_decode_updates

wesketchum committed:
some big cleanup/updates on DAPHNE decoder

wesketchum commented:
OK, this has some fairly extensive cleanup on the DAPHNE decoder, making use of the common unpacking utilities.

To recreate a testing area, you can use `dbtx-create-workarea-from-recipe.py /nfs/home/wketchum/dunedaq_builds/dev_fddaq-v5.3.2-dqm/wketchum_daphne_decoder_update_v5_3_3.yaml`.

After building, to run a test, you can do:

```
(dbt) [wketchum@np04-srv-004 dev_fddaq-v5.3.2-dqm]$ daphne_decoder.py --help
Usage: daphne_decoder.py [OPTIONS] FILENAME

Options:
  --det TEXT                    Subdetector string (default: HD_PDS)
  -n, --nrecords INTEGER        How many Trigger Records to process (default:
                                all)
  --nskip INTEGER               How many Trigger Records to skip (default: 0)
  --channel-map TEXT            Channel map to load (default: None)
  --adc_stats                   Print adc stats (works for streaming data)
  --print-wvfm-samples INTEGER  How many samples in each waveform to print.
  --print-tp-info               Print TP info from DAPHNFrame.
  --help                        Show this message and exit.
  ```
  
  And so, using the file `/nfs/rscratch/wketchum/np02vd_raw_run037233_0000_df-s04-d0_dw_0_20250711T100925.hdf5`, do things like:
  ```
  daphne_decoder.py -n --channel-map SimplePDSChannelMap --adc_stats --print-wvfm-samples 10 --print-tp-info --det VD_Membrane_PDS /nfs/rscratch/wketchum/np02vd_raw_run037233_0000_df-s04-d0_dw_0_20250711T100925.hdf5
  ```
  or do VD_Cathode_PDS.
  
  (Note if you have the recent detdataformat updates, then it would be "VD_MembranePDS" and "VD_CathodePDS")

bieryAtFnal commented:
Writing down my observations as I go (some of which I may need to retract or modify as I learn more)...

- I wonder where the `--print_frame_timestamps` option went.  Maybe it has been renamed to `--print-wvfm-samples`?
- I started out by running both the old and new `daphne_decoder` script on an HDF5 raw data file that was created by the DAPHNE_TPG configuration in the `readout_type_scan` regression test (this file has emulated DAPHNE self-triggered data)
    - the detector type of the DAPHNE fragments produced in that test were `HD_PDS` (as found by running `HDF5LIBS_TestDumpRecord`)
    - running `daphne_decoder.py --det HD_PDS -n 1 <filename>` with the old version produced output along the lines of what I've seen before (which I've had some reservations about, but there weren't any error messages)
    - running `daphne_decoder.py --det HD_PDS -n 1 <filename>` with the new version produced a traceback that ended with the following:
        -   `if 'TPC' in channel_map:`
        -  `TypeError: argument of type 'NoneType' is not iterable`
    - adding a value for the `--channel-map` seemed to help.  Is that now a required parameter?
    - the output from the new version and the command-line options mentioned above has several time printouts that look like they are supposed to be human-readable times, but the python function wasn't called:  for example, `trigger_time=datetime.datetime(2025, 7, 16, 18, 12, 51, tzinfo=<UTC>),`
    - also, the output talks about TDE frames?  `Found 10 TDE frames in this fragment` Should that be PDS or DAPHNE?
    - the new output shows information about the headers, which is nice (and new, compared to the old version)
    - the new output does not show any information about the channel numbers that are present in the data, whereas the older version did.  Nor does it give information about whether the data is from Streaming or Self-Triggered mode.  Maybe this is 'by-design'...
    - there is a typo in both the old and new version, `Processing fnished` (assuming that fnished is not just an abbreviation)

... to be continued ...

bieryAtFnal commented:
- adding the `--adc_stats` command-line option...
    - the old `daphne_decoder` doesn't show anything related to this option, which was a known feature of the old script and self-triggered data
    - the new script shows information for each channel in each fragment, although I wonder if Python functions are being printed out without being evaluated, for example `np.float64(9378.4384765625)`
    - this is a minor request, but could the option be renamed to `adc-stats` (hyphen instead of dash)

- trying out the `--print-wvfm-samples` option...
    - I'm guessing that `--print-wvfm-samples 1` is equivalent to the old `--print_frame_timestamps` option.  That is great, but it would be nice if the new script could print out a human-readable timestamp in addition to the DUNE timestamp.  Maybe that could just be done for the first, overall timestamp.
    - the exponential format for some of the timestamp printouts may lose information for a user who is interested in the fine details (e.g. `ts=1.095430982037824e+17`)

bieryAtFnal commented:
- printing out the TP info with the `--print-tp-info` option looks good to me
- it looks like the `--summary` and ``--check_ts` options have been removed, which I presume is 'by-design'

bieryAtFnal commented:
A couple additional notes from running on a DAPHNE_Streaming file produced by the relevant config in `readout_type_scan.py`:

- when I tried `daphne_decoder.py --print-wvfm-samples 1 --channel-map SimplePDSChannelMap --det HD_PDS -n 1 <filename>`, I saw the following error:
    - `UnboundLocalError: local variable 'pds_ana' referenced before assignment`

- the default output when the script is run on Streaming data includes the printout of the `DAPHNEStreamHeaderData`, but there doesn't seem to be an analogous printout for self-triggered data.  Is correspondng information not available for self-triggered data?

bieryAtFnal commented:
Another note about using the `--print-tp-info` option on a file with self-triggered data...

- it looks like the timestamp of the peak is always 64, but my previous experience is that this value varies somewhat.

e.g.
```
--PRINTING TP INFO--
	Analyzing Frame 0: TS=109543098203773237 DAPHNE_CH=7
		Found 1 TPs:
			TP Peak 0 at ts=64,  adc_integral=47023, adc_max=394,  t_over_baseline=317,  n_subpeaks=2
	Analyzing Frame 1: TS=109543098203774261 DAPHNE_CH=10
		Found 1 TPs:
			TP Peak 0 at ts=64,  adc_integral=129756, adc_max=955,  t_over_baseline=369,  n_subpeaks=2
	Analyzing Frame 2: TS=109543098203775285 DAPHNE_CH=10
		Found 1 TPs:
			TP Peak 0 at ts=64,  adc_integral=21706, adc_max=200,  t_over_baseline=212,  n_subpeaks=1
	Analyzing Frame 3: TS=109543098203776309 DAPHNE_CH=30
		Found 1 TPs:
			TP Peak 0 at ts=64,  adc_integral=30, adc_max=10,  t_over_baseline=22,  n_subpeaks=1
```

wesketchum commented:
Comments/replies on the first batch.
>     * I wonder where the `--print_frame_timestamps` option went.  Maybe it has been renamed to `--print-wvfm-samples`?
Yes, I removed it, in favor of that ... which is not the same thing, but achieves much of the same. I can bring it back if desired (and yes, should have communicated that ...)

>       * running `daphne_decoder.py --det HD_PDS -n 1 <filename>` with the new version produced a traceback that ended with the following:
>         
>         * `if 'TPC' in channel_map:`
>         * `TypeError: argument of type 'NoneType' is not iterable`
>       * adding a value for the `--channel-map` seemed to help.  Is that now a required parameter?

OK, that's a bug that has (now, I hope) been fixed. One _should_ add a channel-map most likely, but it should work without it. Hopefully that's the case with the updated fix.

>       * the output from the new version and the command-line options mentioned above has several time printouts that look like they are supposed to be human-readable times, but the python function wasn't called:  for example, `trigger_time=datetime.datetime(2025, 7, 16, 18, 12, 51, tzinfo=<UTC>),`

It'd be better if they were more human-readable. It was on my todo list to make all the rawdatautils dataclasses have custom string representations that would look nicer...I'd like to put that to a future commit.

>       * also, the output talks about TDE frames?  `Found 10 TDE frames in this fragment` Should that be PDS or DAPHNE?
...you can tell where I copied code from ... fixed. Thank you.

>       * the new output does not show any information about the channel numbers that are present in the data, whereas the older version did.  Nor does it give information about whether the data is from Streaming or Self-Triggered mode.  Maybe this is 'by-design'...

OK, channel number information should be printed when the adc-stats or waveform info is printed, so that's sorta by design. I've added something in the message saying how many frames there are to say what kinds of frames (fragment type) they are (which to do most nicely requires adding python bindings to daqdataformats...PR to come for that).

>       * there is a typo in both the old and new version, `Processing fnished` (assuming that fnished is not just an abbreviation)

now fixed. thank you...

wesketchum commented:
> * adding the `--adc_stats` command-line option...
> ...
>       * the new script shows information for each channel in each fragment, although I wonder if Python functions are being printed out without being evaluated, for example `np.float64(9378.4384765625)`
Yeah, it's a side-effect of the printout on the dataclasses. Again, I'd prefer to handle this by making cleaner/nicer __str__ overrides in the dataclasses, and will look to do that in a future iteration.

>       * this is a minor request, but could the option be renamed to `adc-stats` (hyphen instead of dash)
Glad you agree, as I wanted to do that. Done.

>     * trying out the `--print-wvfm-samples` option...
>       
>       * I'm guessing that `--print-wvfm-samples 1` is equivalent to the old `--print_frame_timestamps` option.  That is great, but it would be nice if the new script could print out a human-readable timestamp in addition to the DUNE timestamp.  Maybe that could just be done for the first, overall timestamp.

Now done (just for the first timestamp)

>       * the exponential format for some of the timestamp printouts may lose information for a user who is interested in the fine details (e.g. `ts=1.095430982037824e+17`)

Indeed. Done.

wesketchum commented:
> A couple additional notes from running on a DAPHNE_Streaming file produced by the relevant config in `readout_type_scan.py`:
> 
>     * when I tried `daphne_decoder.py --print-wvfm-samples 1 --channel-map SimplePDSChannelMap --det HD_PDS -n 1 <filename>`, I saw the following error:
>       
>       * `UnboundLocalError: local variable 'pds_ana' referenced before assignment`
Should now be fixed. Thank you ...

>     * the default output when the script is run on Streaming data includes the printout of the `DAPHNEStreamHeaderData`, but there doesn't seem to be an analogous printout for self-triggered data.  Is correspondng information not available for self-triggered data?
There's not a DAPHNEFrame header data object implemented, as most of the information is just in the 'analysis' data object. So, for now, that's expected, but perhaps creating a different header piece there would be useful ... though I may defer that to when the DAPHNEEthFrame is done and as TP information properly in its header (not trailer).

wesketchum commented:
> Another note about using the `--print-tp-info` option on a file with self-triggered data...
> 
>     * it looks like the timestamp of the peak is always 64, but my previous experience is that this value varies somewhat.
> 
> 
> e.g.
> 
> ```
> --PRINTING TP INFO--
> 	Analyzing Frame 0: TS=109543098203773237 DAPHNE_CH=7
> 		Found 1 TPs:
> 			TP Peak 0 at ts=64,  adc_integral=47023, adc_max=394,  t_over_baseline=317,  n_subpeaks=2
> 	Analyzing Frame 1: TS=109543098203774261 DAPHNE_CH=10
> 		Found 1 TPs:
> 			TP Peak 0 at ts=64,  adc_integral=129756, adc_max=955,  t_over_baseline=369,  n_subpeaks=2
> 	Analyzing Frame 2: TS=109543098203775285 DAPHNE_CH=10
> 		Found 1 TPs:
> 			TP Peak 0 at ts=64,  adc_integral=21706, adc_max=200,  t_over_baseline=212,  n_subpeaks=1
> 	Analyzing Frame 3: TS=109543098203776309 DAPHNE_CH=30
> 		Found 1 TPs:
> 			TP Peak 0 at ts=64,  adc_integral=30, adc_max=10,  t_over_baseline=22,  n_subpeaks=1
> ```

Generally the TP peak 0 should be at ts 64, as that's how the DAPHNEFrame is constructed. In some of the VD PDS files, I do see additional peaks, so I think the fact you don't see any is just a consequence of the input data you are using.

wesketchum committed:
updates/fixes with daphne_decoder based on comments

wesketchum commented:
Note: to run /test this now, one should include https://github.com/DUNE-DAQ/daqdataformats/pull/69. yaml recipe file referenced above has been updated.

bieryAtFnal commented:
Thanks for the fixes and information.

Working backward through our lists...

- part of my confusion about the TP Peak value of 64 was that I recalled seeing values less than 64 in tests that I ran a few weeks ago.  But, I see now that I was looking at `sample_max_0` instead of `sample_start(0)`.  So, I realize now that I was looking at something different.  Would it be worthwhile to add the word "start" to the TP printout, e.g. `TP Peak 0 at start ts=64`, or would that be superfluous because everyone knows that the start time is the interesting timestamp?
- thanks for the information that there isn't currently a DAPHNEFrame header data object implemented.
- thanks for the fix when using `--print-wvfm-samples` on streaming data
- thanks for the human-readable time strings and the non-exponential values, and for the hyphen in the `--adc-stats` option
- thanks for the other fixes and explanations.  I'm fine with `--print_frame_timestamps` being replaced with `--print-wvfm-samples`.  And, it's fine with me to defer better formatting of `window_end_time=datetime.datetime(2025, 7, 17, 14, 51, 19, tzinfo=<UTC>)` until a later commit.

bieryAtFnal reviewed (approved):
LGTM
